op-service/sources/receipts.go

```diff
@@ -4,18 +4,206 @@ import (
 	"context"
 	"fmt"
 	"io"
-	"sync"
+	"time"
 
 	"github.com/ethereum-optimism/optimism/op-service/sources/batching"
+	"github.com/ethereum-optimism/optimism/op-service/sources/caching"
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/common/hexutil"
 	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
 	"github.com/ethereum/go-ethereum/rpc"
 	"github.com/ethereum/go-ethereum/trie"
 
 	"github.com/ethereum-optimism/optimism/op-service/eth"
 )
 
+type ReceiptsProvider interface {
+	// FetchReceipts returns a block info and all of the receipts associated with transactions in the block.
+	// It verifies the receipt hash in the block header against the receipt hash of the fetched receipts
+	// to ensure that the execution engine did not fail to return any receipts.
+	FetchReceipts(ctx context.Context, block eth.BlockID, txHashes []common.Hash) (types.Receipts, error)
+}
+
+type CachingReceiptsProvider struct {
+	inner ReceiptsProvider
+	cache *caching.LRUCache[common.Hash, types.Receipts]
+}
+
+func NewCachingReceiptsProvider(inner ReceiptsProvider, m caching.Metrics, cacheSize int) *CachingReceiptsProvider {
+	return &CachingReceiptsProvider{
+		inner: inner,
+		cache: caching.NewLRUCache[common.Hash, types.Receipts](m, "receipts", cacheSize),
+	}
+}
+
+func (p *CachingReceiptsProvider) FetchReceipts(ctx context.Context, block eth.BlockID, txHashes []common.Hash) (types.Receipts, error) {
+	if r, ok := p.cache.Get(block.Hash); ok {
+		return r, nil
+	}
+
+	r, err := p.inner.FetchReceipts(ctx, block, txHashes)
+	if err != nil {
+		return nil, err
+	}
+
+	p.cache.Add(block.Hash, r)
+	return r, nil
+}
+
+// CachedReceipts provides direct access to the underlying receipts cache.
+// It returns the receipts and true for a cache hit, and nil, false otherwise.
+func (p *CachingReceiptsProvider) CachedReceipts(blockHash common.Hash) (types.Receipts, bool) {
+	return p.cache.Get(blockHash)
+}
+
+type rpcClient interface {
```

Comment:
Depending on whether we want to expose constructors for the `EthClient` to inject any `ReceiptsProvider` implementation, should we make this interface public because we'd then also keep the `RPCReceiptsFetcher` and other implementations public.

Reply:
Keeping the scope of this PR to a refactor, so not doing this now.